scripts/cluster/join.py

```diff
@@ -36,6 +38,23 @@
 cluster_key_file = "{}/cluster.key".format(cluster_dir)
 
 
+def get_fingerprint(addr, port):
+    """
+    Get the certificate fingerprint of the server at addr:port
+
+    :param addr: the address of the server
+    :param port: the port of the server
+    :return: the digest of the server cert
+    """
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```

Comment:
This appears to be creating a second, short-lived, connection, solely for the purpose of getting a peer certificate and checking that the checksum of the certificate matches a fingerprint supplied from another source.

If so, this introduces a race condition -- the connection that would actually be used by the rest of the application has no relationship the certificate that was retrieved in this step, they come from different connections.

I can understand that the deployment model of microk8s may not allow for *actual* live certificates to be used, so a method to identify peers by a fingerprint alone may be useful, but it at least needs to verify the certificate used on the connection in question.

Reply:
You are correct, we create a short-lived connection just to get the certificate.

Python's `requests` package does not allow getting the certificate details. The work we need to do to is described in https://stackoverflow.com/questions/16903528/how-to-get-response-ssl-certificate-from-requests-in-python/52072170#52072170 

I am looking for alternatives before commit to that approach.

Reply:
Discussed in chat, but adding here for reference...

It looks like if you use `streaming=True` as [documented here](https://requests.readthedocs.io/en/master/user/advanced/#body-content-workflow), you can get access to the socket being used by requests and call `getpeercert` on it:

```python
>>> import requests
>>> with requests.get('https://google.com/', stream=True) as resp:
...     print(resp.raw.connection.sock.getpeercert(True))
... 
b'0\x82\x04\xc50\x82\x03\xad\xa0\x03\x02\x01\x02...'
```

Reply:
I am struggling with this and would appreciate any pointers you may be able to offer.

We need to use the same connection to do two things. 1. do the SSL handshake and inspect the host certificate 2. If the certificate fingerprint is acceptable perform a POST request. The closest  I got to this is flow is to keep the tcp connection and perform two HTTP requests similar to [1]. The obstacle I am now facing is that I am asking the connection to not verify the certificates [2] and (I think) this restrict me from getting the SSLsocket form wrap_socket [3] with the error "ssl.SSLError: [SSL: UNKNOWN_PROTOCOL] unknown protocol". Further more when wrapping the socket I suspect the socket gets detached and a new handshake takes place [4].

Any hints you have to offer are much appreciated.

[1] https://stackoverflow.com/questions/2009243/how-to-speed-up-pythons-urllib2-when-doing-multiple-requests
[2] https://docs.python.org/2/library/httplib.html#httplib.HTTPSConnection
[3] https://docs.python.org/3/library/ssl.html#ssl.SSLContext.wrap_socket
[4] https://github.com/python/cpython/blob/3.8/Lib/ssl.py#L1006

Reply:
Spent some more time digging into this. We now use `http.client` manuly force the SSL handshake with connect() and if the certificate checks out we do the POST request.